### ISSUE TITLE
Poly loses modulus after arithemetic

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#13545]
```
>>> Poly(x + 1, x, modulus=2) + 1
Poly(x + 2, x, domain='ZZ')
>>> Poly(x + 1, x, modulus=2) + Poly(1, x, modulus=2)
Poly(x, x, modulus=2)
```

I would expect the first to give the same answer as the second.

**Comment by jksuom** [comment]
I agree. Coefficient unification should be amended. It is currently defined strictly by size, the bigger domain being taken as the common domain. From domains.py:

        Known domains (from smallest to largest):

        - ``GF(p)``
        - ``ZZ``
        - ``QQ``
        - ``RR(prec, tol)``
        - ``CC(prec, tol)``
        - ``ALG(a, b, c)``
        - ``K[x, y, z]``
        - ``K(x, y, z)``
        - ``EX``

Two domains, `A` and `B`, should be unified to a domain `C` only if there are homomorphisms `A -> C` and `B -> C`. If one of the domains has characteristic `p != 0` then also the unified domain `C` must have characteristic `p`. In particular, the pair `ZZ`, `GF(p)` should be unified to `GF(p)` since there is a natural homomorphism `ZZ -> GF(p)`. Also the pairs `ZZ[x]`, `GF(p)` and `ZZ`, `GF(p)[x]` can be unified to `GF(p)[x]` (and similarly for fraction fields).

On the other hand, there are no homomorphisms from fields of characteristic zero (`QQ`, ...) to domains containing `GF(p)`, and therefore the unification should fail in those cases.

**Comment by asmeurer** [comment]
That sounds reasonable. It also implies that fields of different characteristics should not unify, which is apparently not the case currently:

```
>>> Poly(x, modulus=2) + Poly(x, modulus=3)
Poly(-x, x, modulus=3)
```

We also probably really want only natural homomorphisms, at least for automatic unification (this maybe could be an issue for fields of order p^n? We don't have those yet, though, so it isn't a concern yet)

This all sounds quite similar to type unification https://en.wikipedia.org/wiki/Unification_(computer_science), except with homomorphisms instead of substitutions.

**Comment by guptapratykshh** [comment]
I think so this is resolved as it returns `Poly(x, x, modulus=2)` .  Please look into this. 

```
In [13]: Poly(x + 1, x, modulus=2) + 1
Out[13]: Poly(x, x, modulus=2)

In [14]: Poly(x + 1, x, modulus=2) + Poly(1, x, modulus=2)
Out[14]: Poly(x, x, modulus=2)
```
